### PR TITLE
Support external window usage updates

### PR DESCRIPTION
**Pull request by jsantos17**
This is the first piece required to support [Facebook rate limiting](https://developers.facebook.com/docs/graph-api/overview/rate-limiting/).

The main difference is that Facebook will provide us with a limit rather so we don't need to count the requests/events we make. This will be followed with another PR implementing the Facebook-specific code in the URL datasource.

[ch13331]

### CODE REVIEW

File: foundation/src/main/scala/quasar/RateLimiter.scala [hunk 1]
```diff
@@ -64,34 +64,38 @@ final class RateLimiter[F[_]: Concurrent: Timer, A: Hash] private () {
...
     } yield {
       RateLimiterEffects[F](
-        limit(config, stateRef),
-        backoff(config, stateRef))
+        limit(config, stateRef, setUsage),
+        backoff(config, stateRef),
+        setWindowUsage(stateRef))
     }
 
+  private def setWindowUsage(stateRef: Ref[F, State[F]])(usage: Int)
```
Comment (jsantos17) [inline foundation/src/main/scala/quasar/RateLimiter.scala:72]: This almost the same as `backoff`, but allows specifying the limit. The URL datasource will call this on every request that includes usage information (every request for Facebook).

**Comment by jsantos17** [comment]
🛑 because I haven't opened the PR fixing the URL datasource.

**Review by rintcius — COMMENTED** [review]
Nice!

For the Facebook implementation, after the usage has been set to 100%, is there a way to ask only the usage or does this always need a new request? 

Just wondering, because if it needs a new request then I guess we need some kind heuristic to determine how long to back off.

Not directly related to this PR, but I saw there's `RateLimitUpdater` but I don't think this is used, so maybe we can remove this?

File: foundation/src/main/scala/quasar/RateLimiter.scala [hunk 2]
```diff
@@ -131,7 +135,8 @@ final class RateLimiter[F[_]: Concurrent: Timer, A: Hash] private () {
    *
    * If the queue is not empty, we enqueue the new request.
    */
-  private def limit(config: RateLimiterConfig, stateRef: Ref[F, State[F]]): F[Unit] = {
+  private def limit(config: RateLimiterConfig, stateRef: Ref[F, State[F]], setUsage: Int => Int)
```
[the remark this answers is not in the document]
Comment (jsantos17) [reply foundation/src/main/scala/quasar/RateLimiter.scala:138]: Definitely. Didn't realy they had the same name. Will fix.

**Comment by jsantos17** [comment]
@rintcius 

Unfortunately there's not a way to only check the current limits. However, Facebook does provide time windows for each API and we still need them configured in the URL datasource for external rate limiting. From my experimenting, it's safe to resume requests after waiting for the window to end.

For instance the Marketing API and the Instagram Basic API both have windows of 1 hour and after hitting 100% usage, usage will go down to 50% after 1 hour. It goes down to zero after ~1.5 hours. Might need some tweaking when configuring the datasource, since the documentation doesn't really match actual behavior.

**Review by alissapajer — APPROVED** [review]

**Review by rintcius — APPROVED** [review]

**Pull request merged**